### PR TITLE
HTBHF-2656 Expire claim in pending expiry 

### PR DESCRIPTION
**Pull request by dwybourn**
An eligibility check that returns ineligible will expire a claim if that claim has been pending expiry for 16 or more weeks.

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java [hunk 1]
```diff
@@ -17,31 +17,49 @@
...
+    private final EligibilityAndEntitlementService eligibilityAndEntitlementService;
+    private final MessageContextLoader messageContextLoader;
+    private final PaymentCycleService paymentCycleService;
+    private final MessageQueueClient messageQueueClient;
+    private final EligibilityDecisionHandler eligibilityDecisionHandler;
+    private final PregnancyEntitlementCalculator pregnancyEntitlementCalculator;
+    private final ClaimMessageSender claimMessageSender;
+
+    public DetermineEntitlementMessageProcessor(
+            @Value("${payment-cycle.maximum-pending-expiry-duration}") Period maximumPendingExpiryDuration,
```
Comment (dwybourn) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java]: This could do with some refactoring in it's own PR. Would be worth pairing on.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java [hunk 2]
```diff
@@ -82,8 +104,16 @@ private void handleDecision(Claim claim, PaymentCycle previousPaymentCycle, Paym
...
+        } else if (claimHasBeenPendingExpiryForLongerThanTheMaximumDuration(claim, messageTimestamp)) {
+            claim.updateClaimStatus(EXPIRED);
+            claimMessageSender.sendReportClaimMessage(claim, decision.getDateOfBirthOfChildren(), UPDATED_FROM_PENDING_EXPIRY_TO_EXPIRED);
         }
-        //TODO HTBHF-1296: If not ACTIVE, PENDING_EXPIRY will be moved to EXPIRED after 16 weeks.
+    }
+
+    private boolean claimHasBeenPendingExpiryForLongerThanTheMaximumDuration(Claim claim, LocalDateTime messageTimestamp) {
+        // !isAfter has the same effect as (isBefore || isEqual). Unfortunately isBeforeOrEqual does not exist on LocalDateTime.
+        return !claim.getClaimStatusTimestamp().isAfter(messageTimestamp.minus(maximumPendingExpiryDuration))
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java]: I'd suggest swapping the if statement around so that checking for PENDING_EXPIRY comes first.

Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java]: Swapped around.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java [hunk 3]
```diff
@@ -82,8 +104,16 @@ private void handleDecision(Claim claim, PaymentCycle previousPaymentCycle, Paym
             }
         } else if (claim.getClaimStatus() == ACTIVE) {
             eligibilityDecisionHandler.handleIneligibleDecision(claim, previousPaymentCycle, currentPaymentCycle, decision);
+        } else if (claimHasBeenPendingExpiryForLongerThanTheMaximumDuration(claim, messageTimestamp)) {
+            claim.updateClaimStatus(EXPIRED);
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java]: This needs an explicit `save(claim)`. Sadly I think this might need another component added as we don't seem to have the `claimRepository` injected in this class.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessorTest.java [hunk 4]
```diff
@@ -162,6 +176,78 @@ void shouldSuccessfullyProcessMessageAndHandleIneligibleDecision() {
...
+        MessageStatus messageStatus = processor.processMessage(message);
+
+        //Then
+        assertThat(messageStatus).isEqualTo(COMPLETED);
+        assertThat(context.getClaim().getClaimStatus()).isEqualTo(PENDING_EXPIRY);
+        verify(messageContextLoader).loadDetermineEntitlementContext(message);
+        verify(eligibilityAndEntitlementService).evaluateClaimantForPaymentCycle(context.getClaim().getClaimant(),
+                context.getCurrentPaymentCycle().getCycleStartDate(),
+                context.getPreviousPaymentCycle());
+        verifyNoMoreInteractions(claimMessageSender);
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessorTest.java]: SHouldn't this be `verifyNoInteractions`? Can't see any interaction with this mock.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java (hunk 3 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessor.java]: Added explicit save.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessorTest.java (hunk 4 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/message/processor/DetermineEntitlementMessageProcessorTest.java]: Changed to verifyNoInteractions
